### PR TITLE
HPCC-24713 Records in TLK leaf node should not contain payload fields

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Regression suite and dumpkey

### CODE REVIEW

File: thorlcr/activities/indexwrite/thindexwriteslave.cpp [hunk 1]
```diff
@@ -141,26 +140,41 @@ class IndexWriteSlaveActivity : public ProcessSlaveActivity, public ILookAheadSt
...
+        bool isVariable = false;
+        unsigned maxDiskRecordSize;
         IOutputMetaData * diskSize = helper->queryDiskRecordSize();
         //Need to adjust the size if the last field is used in the special fileposition location.
         size32_t fileposSize = hasTrailingFileposition(diskSize->queryTypeInfo()) ? sizeof(offset_t) : 0;
         assertex(!(diskSize->getMetaFlags() & MDFneedserializedisk));
-        if (diskSize->isVariableSize())
+        if (isTlk)
+        {
+            maxDiskRecordSize =  helper->getKeyedSize();
```
Comment (jakesmith) [inline thorlcr/activities/indexwrite/thindexwriteslave.cpp:159]: trivial/formatting: double space after =

File: thorlcr/activities/indexwrite/thindexwriteslave.cpp [hunk 2]
```diff
@@ -452,6 +466,13 @@ class IndexWriteSlaveActivity : public ProcessSlaveActivity, public ILookAheadSt
 
                     CMessageBuffer msg;
                     MemoryAttr dummyRow;
+                    size32_t keyedSize = helper->getKeyedSize();
+                    if (keyedSize == (unsigned) -1)
+                    {
+                        // For some reason, if there is no payload, getKeyedSize returns -1
+                        const RtlRecord &indexRecord = helper->queryDiskRecordSize()->queryRecordAccessor(true);
+                        keyedSize = indexRecord.getFixedOffset(indexRecord.getNumKeyedFields());
+                    }
```
Comment (jakesmith) [inline thorlcr/activities/indexwrite/thindexwriteslave.cpp:475]: perhaps this block should move to start() as same as block in open(), and record keyedSize as member and use here and in open()

**Review by jakesmith — APPROVED** [review]
@richardkchapman - looks good, 1 question/suggestion.

**Comment by richardkchapman** [comment]
Build failed because of a rebase hiccup. Will fix and repush

**Review by jakesmith — APPROVED** [review]
@richardkchapman - looks good afaics

**Review by ghalliday — APPROVED** [review]

**Comment by richardkchapman** [comment]
@jakesmith Changed as per suggestions - please re-review

**Comment by richardkchapman** [comment]
@jakesmith Never mind I think you arteady did! Had not refreshed for some reason

**Comment by jakesmith** [comment]
> @jakesmith Never mind I think you already did! Had not refreshed for some reason

yes, I refreshed the review status, then re-approved.

**Comment by jakesmith** [comment]
@richardkchapman - however, I see it's failing some regression tests..

**Review by ghalliday — APPROVED** [review]

File: ecl/hthor/hthorkey.cpp [hunk 3]
```diff
@@ -774,7 +774,7 @@ void CHThorIndexReadActivityBase::verifyIndex(IKeyIndex * idx)
             keySize = idx->keySize();
             //The index rows always have the filepositions appended, but the ecl may not include a field
             unsigned fileposSize = idx->hasSpecialFileposition() && !hasTrailingFileposition(eclKeySize.queryTypeInfo()) ? sizeof(offset_t) : 0;
-            if (keySize != eclKeySize.getFixedSize() + fileposSize)
+            if (keySize != eclKeySize.getFixedSize() + fileposSize && !idx->isTopLevelKey())
```
Comment (jakesmith) [inline ecl/hthor/hthorkey.cpp:777]: personal: I'd prefer some brackets for clarity here.

**Review by jakesmith — APPROVED** [review]
@richardkchapman @ghalliday - looks good - 1 [personal:] comment.

**Pull request merged**